### ISSUE TITLE
Refactor to use TryFrom

### ISSUE THREAD

**Issue by hawkw** [hawkw/sos-kernel#85]
Since the [`TryFrom`](https://doc.rust-lang.org/std/convert/trait.TryFrom.html) trait is in unstable Rust (rust-lang/rust#33417), we should rewrite a lot of failable conversions into `TryFrom` implementations. For example, a _lot_ of the [ELF parsing code](https://github.com/hawkw/sos-kernel/tree/master/elf) that returns `Result<Self, ElfError>` can probably be rewritten to use `TryFrom`.

Where removing existing failable conversion methods would break a lot of other code, we can temporarily preserve the existing API by having it wrap a `try_from` call, and phase it out gradually.

We don't have to worry about `TryFrom` not being on stable, since we can't build SOS on stable anyway.

**Comment by hawkw** [comment]
I'm marking this as "easy" since it is a reasonably simple refactoring that doesn't require a great deal of OS knowledge.

**Comment by tirkarthi** [comment]
As an update to the issue the above RFC was implemented and merged with PR : https://github.com/rust-lang/rust/pull/49305 . It caused notable breakages in the ecosystem and got reverted : https://github.com/rust-lang/rust/pull/49518

**Comment by hawkw** [comment]
@tirkarthi my understanding is that all that was reverted is the addition of `TryFrom` _to the prelude_, the traits are still stabilized, correct?

**Comment by tirkarthi** [comment]
Yes, the team is looking for a way to introduce this with less breakages.
